sam_view.c

```diff
@@ -855,188 +1068,137 @@ int main_samview(int argc, char *argv[])
-                                }
-                                count++;
-                            } else if (settings.unmap) {
-                                b->core.flag |= BAM_FUNMAP;
-                                if (check_sam_write1(out, header, b, fn_out, &ret) < 0) break;
-                            } else {
-                                if (un_out) { if (check_sam_write1(un_out, header, b, fn_un_out, &ret) < 0) break; }
-                            }
-                        }
-                        if (result < -1) {
-                            print_error("view", "retrieval of region %d failed due to truncated file or corrupt BAM index file", iter->curr_tid);
-                            ret = 1;
-                        }
-
-                        hts_itr_multi_destroy(iter);
-                    } else {
-                        fprintf(stderr, "[main_samview] iterator could not be created. Aborting.\n");
-                    }
-                } else {
-                    fprintf(stderr, "[main_samview] region list is empty or could not be created. Aborting.\n");
+        int r;
+        errno = 0;
+        while ((r = sam_read1(settings.in, settings.header, b)) >= 0) { // read one alignment from `in'
+            if (!process_aln(settings.header, b, &settings)) {
+                if (!settings.is_count) {
+                    change_flag(b, &settings);
+                    if (check_sam_write1(settings.out, settings.header, b, settings.fn_out, &ret) < 0) break;
                 }
-                hts_idx_destroy(idx); // destroy the BAM index
+                settings.count++;
+            } else if (settings.unmap) {
+                b->core.flag |= BAM_FUNMAP;
+                if (check_sam_write1(settings.out, settings.header, b, settings.fn_out, &ret) < 0) break;
             } else {
-                fprintf(stderr, "[main_samview] random alignment retrieval only works for indexed BAM or CRAM files.\n");
+                if (settings.un_out) { if (check_sam_write1(settings.un_out, settings.header, b, settings.fn_un_out, &ret) < 0) break; }
             }
         }
+        if (r < -1) {
+            print_error_errno("view", "error reading file \"%s\"", settings.fn_in);
+            ret = 1;
+        }
         bam_destroy1(b);
-    } else {
-        if ((has_index_file && optind >= argc - 2) || (!has_index_file && optind >= argc - 1)) { // convert/print the entire file
-            bam1_t *b = bam_init1();
-            int r;
-            errno = 0;
-            while ((r = sam_read1(in, header, b)) >= 0) { // read one alignment from `in'
-                if (!process_aln(header, b, &settings)) {
-                    if (!is_count) {
+    }
+    else    // retrieve alignments in specified regions using the generic iterator
+    {
+        int i;
+        bam1_t *b;
+        b = bam_init1();
+
+        for (i = (has_index_file)? optind+2 : optind+1; i < argc; ++i) {
+            int result;
```

Comment:
I think it should be possible to get rid of most of this version by making it call `multi_region_init()` followed by `multi_region_view()` for each region separately.  If it works it would be a very useful simplification.